src/Planner/PlannerJoins.cpp

```diff
@@ -357,24 +349,58 @@ void buildJoinClause(
                 join_clause.addCondition(left_expression_side, node);
             }
         }
+        else
+        {
+            auto support_mixed_join_condition = planner_context->getQueryContext()->getSettingsRef().allow_experimental_join_condition;
+            if (support_mixed_join_condition)
+            {
+                /// expression involves both tables.
+                /// `expr1(left.col1, right.col2) == expr2(left.col3, right.col4)`
+                const auto * node = appendExpression(mixed_dag, join_expression, planner_context, join_node);
```

Comment:
I found that dag can be incorrect with `join_use_nulls` or when column change thier type, e.g. `t0.x :: Int8` and `t0.x :: UInt8` and expression is `JOIN ON t0.x = t1.x AND t0.x + t1.x > 2`. It happens because here we have original types in columns in `join_expression` as if it supposed to be executed before joining. But we update tyypes of columns for mentioned cases right after columns goes into join, so inside HashJoin `mixed_dag` is executed on top of nullable or casted to common type columns, what leads to logical error.
I tried to fix it, but haven't come up with a proper solution yet. Maybe for such case we need to preserve original columns and pass it into join additionally with changed names and this dag should use that columns.
Or handle it here to build a dag that is aware about type changes.
Or just forbid such cases with `NOT_IMPLEMENTED` for now.

Reply:
> What if we add code handling mixedExpression into joinRightColumns under if (mixed_join_expression) (with > boolean template flag and if constexpr if it would be some overhead). Is it possible?

Do you mean ?
```c++
template<bool has_mixed_join_expression >
size_t joinRightColumns()
{
    if constexpr(has_mixed_join_expression)
    {
        // new way to handle
    }
    else
    {
       // as before
    }
}
```

Reply:
😂`join_use_nulls` is used in our case.

Reply:
it's strange, this case does not fail.
```
0.x :: Int8 and t0.x :: UInt8 and expression is JOIN ON t0.x = t1.x AND t0.x + t1.x > 2
```

Reply:
Could we add a `IQueryTreePass` to rewrite the query tree before enter `buildJoinClause` ?

Reply:
> I found that dag can be incorrect with `join_use_nulls` or when column change thier type, e.g. `t0.x :: Int8` and `t0.x :: UInt8` and expression is `JOIN ON t0.x = t1.x AND t0.x + t1.x > 2`. It happens because here we have original types in columns in `join_expression` as if it supposed to be executed before joining. But we update tyypes of columns for mentioned cases right after columns goes into join, so inside HashJoin `mixed_dag` is executed on top of nullable or casted to common type columns, what leads to logical error. I tried to fix it, but haven't come up with a proper solution yet. Maybe for such case we need to preserve original columns and pass it into join additionally with changed names and this dag should use that columns. Or handle it here to build a dag that is aware about type changes. Or just forbid such cases with `NOT_IMPLEMENTED` for now.

yes, that's a big issue, we have a try.

Reply:
@lgbo-ustc 

> > What if we add code handling mixedExpression into joinRightColumns under if (mixed_join_expression) (with > boolean template flag and if constexpr if it would be some overhead). Is it possible?
> 
> Do you mean ? ...


Only if amount of common code that would be common is significant and `handle mixed condition` is not too big, like. I'm not sure if it's feasible or not, though
 
```c++
template<bool has_mixed_join_expression >
size_t joinRightColumns()
{
/// common code
    if constexpr(has_mixed_join_expression)
    {
        // handle mixed condition
    }
/// common code
}
```

Reply:
About issue with `join_use_nulls` not sure if it's doable with a pass or not. probably we may have have a pass that change types in such expressions in `JOIN ON`, seems more like workaround, but I don't have an idea hot to fix it properly for the moment.

Reply:
I have had some idea about the `query pass`, I think could make it in few days, just take a try at first.

Reply:
About the common code in `joinRightColumns`, I have no more idea at present. If you have time, could you help to think about it?